### ISSUE TITLE
[Synology NAS] cross compile from source fails

### ISSUE THREAD

**Issue by hgy59** [navidrome/navidrome#1475]
### Description

navidrome fails to build for Synology Diskstations with https://github.com/SynoCommunity/spksrc/
see https://github.com/SynoCommunity/spksrc/issues/4973#issuecomment-974629297

### Platform information

 - Navidrome version: v0.47.0 (tried also 0.45.1, 0.44.1, 0.42.1)

### Additional information
 - the frontend builds successfully with `make setup` and `make buildjs` (with nodejs v16)
 - the dev environment has installed taglib-dev v1.11.1 (tried also with taglib v1.12)
 - `make build` fails with:
   ```
   cd /spksrc/cross/navidrome/work-x64-6.1/navidrome-0.47.0 && env PARALLEL_MAKE=max GOPATH=/spksrc/cross/navidrome/../../distrib/go CGO_ENABLED=0 PATH=/spksrc/cross/navidrome/work-x64-6.1/../../../native/go/work-native/go/bin/:$PATH GOARCH=amd64 GOOS=linux PKG_CONFIG_LIBDIR=/spksrc/cross/navidrome/work-x64-6.1/install//usr/local/navidrome/lib/pkgconfig WORK_DIR=/spksrc/cross/navidrome/work-x64-6.1 INSTALL_PREFIX=/usr/local/navidrome TC=syno-x64-6.1  LD="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld" LDSHARED="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-gcc -shared" CPP="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-cpp" NM="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-nm" CC="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-gcc" AS="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-as" RANLIB="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ranlib" CXX="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-g++" AR="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ar" STRIP="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-strip" OBJDUMP="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-objdump" READELF="/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-readelf" CFLAGS="-I/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/x86_64-pc-linux-gnu/sys-root/usr/include -I/spksrc/cross/navidrome/work-x64-6.1/install//usr/local/navidrome/include " CPPFLAGS="-I/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/x86_64-pc-linux-gnu/sys-root/usr/include -I/spksrc/cross/navidrome/work-x64-6.1/install//usr/local/navidrome/include " CXXFLAGS="-I/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/x86_64-pc-linux-gnu/sys-root/usr/include -I/spksrc/cross/navidrome/work-x64-6.1/install//usr/local/navidrome/include " LDFLAGS="-L/spksrc/toolchain/syno-x64-6.1/work/x86_64-pc-linux-gnu/x86_64-pc-linux-gnu/sys-root/lib -L/spksrc/cross/navidrome/work-x64-6.1/install//usr/local/navidrome/lib -Wl,--rpath-link,/spksrc/cross/navidrome/work-x64-6.1/install//usr/local/navidrome/lib -Wl,--rpath,/usr/local/navidrome/lib " make build
   make[2]: Entering directory '/spksrc/cross/navidrome/work-x64-6.1/navidrome-0.47.0'
   WARNING: This command does not build the frontend, it uses the latest built with 'make buildjs'
   go build -ldflags="-X github.com/navidrome/navidrome/consts.gitSha=source_archive -X github.com/navidrome/navidrome/consts.gitTag=v0.47.0-SNAPSHOT" -tags=netgo
   # github.com/navidrome/navidrome/scanner/metadata/taglib
   scanner/metadata/taglib/taglib.go:22:15: undefined: Read
   scanner/metadata/taglib/taglib.go:35:16: assignment mismatch: 2 variables but 1 value
   scanner/metadata/taglib/taglib.go:44:20: assignment mismatch: 2 variables but 1 value
   make[2]: *** [Makefile:73: build] Error 2
   ```

- with `make single` any cross compilation fails with ~ `unsupported option -m64`
- for x86_64 I got a successful build, but `navidrome --version` reports `0.0.0-SNAPSHOT (none)`
- I will prefer the `make build` method but need help on cross compiling options...

Any hint on configuring CGO or building navidrome without taglib would be appreciated.

**Comment by deluan** [comment]
> with make single any cross compilation fails with ~ unsupported option -m64

What GOOS/GOARCH are you using with `make single`? It should work as long as your docker is properly installed

> Any hint on configuring CGO

`CGO_ENABLED` **must** be 1, or else you won't be able to generate a working build. Navidrome depends on SQLite, which requires CGO to be enabled

What error do you get when `CGO_ENABLED=1` ?

**Comment by bardo** [comment]
Similar issue on SmartOS (Illumos-based) with the latest trunk:
```
$ export GOOS=illumos  # This is needed because of a SmartOS packaging issue which defaults to Solaris
$ make build
WARNING: This command does not build the frontend, it uses the latest built with 'make buildjs'
go build -ldflags="-X github.com/navidrome/navidrome/consts.gitSha=67f2a89d -X github.com/navidrome/navidrome/consts.gitTag=v0.47.5-SNAPSHOT" -tags=netgo
# github.com/navidrome/navidrome/scanner/metadata/taglib
scanner/metadata/taglib/taglib.go:22:15: undefined: Read
make: *** [Makefile:73: build] Error 2
```

Different error with `CGO_ENABLED=1` though:
```
$ export GOOS=illumos CGO_ENABLED=1
$ make build
WARNING: This command does not build the frontend, it uses the latest built with 'make buildjs'
go build -ldflags="-X github.com/navidrome/navidrome/consts.gitSha=67f2a89d -X github.com/navidrome/navidrome/consts.gitTag=v0.47.5-SNAPSHOT" -tags=netgo
# github.com/navidrome/navidrome
/opt/local/go118/pkg/tool/solaris_amd64/link: running g++ failed: exit status 1
Undefined                       first referenced
 symbol                             in file
sendfile                            /tmp/go-link-3055384394/go.o
ld: fatal: symbol referencing errors. No output written to $WORK/b001/exe/a.out
collect2: error: ld returned 1 exit status

make: *** [Makefile:73: build] Error 2
```

The installed `taglib` version is at 1.12:
```
$ pkgin ls | grep taglib
taglib-1.12          Library for accessing id tags in various media files
```

**Comment by hgy59** [comment]
As @publicarray suggested in https://github.com/SynoCommunity/spksrc/issues/4973#issuecomment-1006535726 navidrome successfully builds with the toolchains of DSM 7. Those have a newer gcc (version 7.5 vs 4.9).

navidrome 0.47.5 successfully builds with:
- gcc 7.5
- tablib 1.12 (fails with taglib 1.11.1)
- go 1.18.4
- node 16.15.1

**Comment by hgy59** [comment]
@bardo what compiler version do you use?

**Comment by bardo** [comment]
@hgy59 GCC 9.3.0:

```
$ gcc --version
gcc (GCC) 9.3.0
```

**Comment by hgy59** [comment]
Finally succeeded to build navidrome even for toolchains with gcc 4.9.x.
But needed to patch taglib_wrapper.cpp as those old gcc are missing full c++11 compatibility.

```patch
# fix for older gcc (like 4.9.x)
# - mp3 tags: use explicit type for mp3File->ID3v2Tag()->frameListMap()
#             as older gcc cannot dereference std::initializer_list correctly
# - mp4 tags: use mp4File->tag()->item instead of mp4File->tag()->itemMap() to access covers
#             as side effect this fixes the build with older gcc
#
--- scanner/metadata/taglib/taglib_wrapper.cpp.orig	2022-01-22 01:35:19.000000000 +0100
+++ scanner/metadata/taglib/taglib_wrapper.cpp	2022-07-30 13:44:25.711760171 +0200
@@ -95,7 +95,7 @@
   if (TagLib::MPEG::File *
       mp3File{dynamic_cast<TagLib::MPEG::File *>(f.file())}) {
     if (mp3File->ID3v2Tag()) {
-      const auto &frameListMap{mp3File->ID3v2Tag()->frameListMap()};
+      const TagLib::ID3v2::FrameListMap &frameListMap{mp3File->ID3v2Tag()->frameListMap()};
       hasCover = !frameListMap["APIC"].isEmpty();
     }
   }
@@ -107,8 +107,11 @@
   // ----- MP4
   else if (TagLib::MP4::File *
            mp4File{dynamic_cast<TagLib::MP4::File *>(f.file())}) {
-    auto &coverItem{mp4File->tag()->itemMap()["covr"]};
-    TagLib::MP4::CoverArtList coverArtList{coverItem.toCoverArtList()};
+    if (!mp4File->tag()->itemMap().contains("covr")) {
+        return false;
+    }
+    auto const &coverItem = mp4File->tag()->item("covr");
+    auto const coverArtList = coverItem.toCoverArtList();
     hasCover = !coverArtList.isEmpty();
   }
   // ----- Ogg
```

---

BTW
`mp4File->tag()->itemMap()` is marked as deprecated in tablib 1.12
